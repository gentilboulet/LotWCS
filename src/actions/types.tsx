import { IInitialStateAction } from '../actions/initial';
import { IHeaderAction } from '../actions/header';
import { IChiAction } from '../actions/chi';
import { ISkillAction } from '../actions/skills';
import { ILoresheetAction } from '../actions/loresheets';
import { IHistoryAction } from '../actions/history';
import { IKungFuAction } from '../actions/kungfus';
import { IVirtueAction } from '../actions/virtues';

export type IAction =
  IInitialStateAction | IHeaderAction | ISkillAction | IHistoryAction |
  ILoresheetAction | IChiAction | IKungFuAction | IVirtueAction;
