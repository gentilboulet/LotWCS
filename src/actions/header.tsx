import * as constants from '../constants/header';
import { IPerk } from '../types/perks';

import { ranks } from '../data/ranks';
import { IDataRank } from '../types/ranks';
import { archetypes } from '../data/archetypes';
import { IDataArchetype } from '../types/archetypes';

export interface IHeaderSetName {
  type: constants.HEADER_SET_NAME;
  name: string;
}

export interface IHeaderSetConcept {
  type: constants.HEADER_SET_CONCEPT;
  concept: string;
}

export interface IHeaderSetArchetype {
  type: constants.HEADER_SET_ARCHETYPE;
  archetype: string;
}

export interface IHeaderSetRank {
  type: constants.HEADER_SET_RANK;
  rank: string;
  value: number;
  perks: IPerk[];
}

export type IHeaderAction = IHeaderSetName | IHeaderSetConcept | IHeaderSetArchetype | IHeaderSetRank;

export function setName(s: string): IHeaderSetName {
  return {
    type: constants.HEADER_SET_NAME,
    name: s
  };
}

export function setConcept(s: string): IHeaderSetConcept {
  return {
    type: constants.HEADER_SET_CONCEPT,
    concept: s
  };
}

export function setArchetype(inputKey: string): IHeaderSetArchetype {
  const foundArchetype = archetypes.find((archetype: IDataArchetype) => { return archetype.key === inputKey; });
  if (! foundArchetype ) { throw 'Unknown archetype "' + inputKey + '"'; }
  return {
    type: constants.HEADER_SET_ARCHETYPE,
    archetype: inputKey
  };
}

export function setRank(rank: string): IHeaderSetRank {
  const foundRank = ranks.find((rk: IDataRank) => { return rk.key === rank; });
  if (! foundRank ) { throw 'Unknown rank "' + rank + '"'; }
  return {
    type: constants.HEADER_SET_RANK,
    rank: rank,
    value: foundRank.value,
    perks: foundRank.perks as IPerk[],
  };
}
