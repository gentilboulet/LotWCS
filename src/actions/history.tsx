import * as constants from '../constants/history';

export interface IhistoryDeleteUpTo {
  type: constants.HISTORY_DELETE;
  id: number;
}

export type IHistoryAction = IhistoryDeleteUpTo;

export function historyDeleteUpTo(id: number): IhistoryDeleteUpTo {
  return {
    type: constants.HISTORY_DELETE,
    id: id,
  };
}
