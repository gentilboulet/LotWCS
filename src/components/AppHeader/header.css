.header {
  text-align: center;
  background-color: #e4e4e4;
  border: 1px solid lightgrey;
}

.link-list {
  display: flex;
  margin: auto;
  width: 50%;
  justify-content: space-around;
}
