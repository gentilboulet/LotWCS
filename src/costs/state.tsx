import { ICost } from 'costs/types';
import { IDiscount } from 'perks/types/discounts';
import { IStoreState } from 'state/types';

import * as constants from 'perks/constants/discounts';
import * as dataLoresheets from 'data/loresheets';
import * as loresheets from 'state/loresheets';
import * as skills from 'state/skills';

import * as derived from 'state/derived';

// helper functions
function _canHandleCost(state: IStoreState, cost: ICost): boolean {
  if ( (state.get('destiny') - cost.destiny) < 0 ) { return false; }
  if ( (state.get('entanglement') - cost.entanglement) < 0 ) { return false; }
  return true;
}

function _handleDiscount(state: IStoreState, idx: number, cost: number): ICost {
  if (idx === -1) {
    return { destiny: cost, discountIdx: idx, discountNewValue: 0, entanglement: 0, };
  } else {
    const discount = state.getIn(['discounts', idx]).value;
    const remainingCost = Math.max(0, cost - discount);
    const usedDiscountValue = cost - remainingCost;
    return {
      destiny: remainingCost,
      discountIdx: idx,
      discountNewValue: discount - usedDiscountValue,
      entanglement: 0,
    };
  }
}

/*
  Export canBuy and getCost for some actions. Available :
    canBuySkill
    getCostSkill
    canBuySpeciality
    getCostSpeciality
    canOpenLoresheet
    getCostOpenLoresheet
    canBuyOptionLoresheet
    getCostsArrayBuyOptionLoresheet
 */

export function canBuySkill(state: IStoreState, skill: string): boolean {
  const value = skills.getSkillValue(state, skill);
  if ( (value + 5) > derived.maxSkillBonus(state) ) { return false; }
  const cost = getCostSkill(state, skill);
  return _canHandleCost(state, cost);
}

export function getCostSkill(state: IStoreState, skill: string): ICost {
  const defCost = 2;

  const idx = state.get('discounts')
    .findIndex((r: IDiscount) => {
      return ( r.type === constants.DISCOUNT_SKILL )
      && (r.skills.findIndex((s: string) => (s === skill) ) >= 0);
    });

  return _handleDiscount(state, idx, defCost);
}

export function canBuySpeciality(state: IStoreState, skill: string, speciality: string): boolean {
  const specialityIdx = skills.getSpecialityIndex(state, skill, speciality);

  if ( specialityIdx < 0 ) { return false; }

  const cost = getCostSpeciality(state, skill, speciality);
  return _canHandleCost(state, cost);
}

export function getCostSpeciality(state: IStoreState, skill: string, speciality: string): ICost {
  const defCost = 1;

  const idx = state.get('discounts')
    .findIndex((r: IDiscount) => {
      return ( r.type === constants.DISCOUNT_SKILL )
        && (r.skills.findIndex((s: string) => (s === skill)) >= 0);
    });

  return _handleDiscount(state, idx, defCost);
}

export function canOpenLoresheet(state: IStoreState, uid: string, openCost: number): boolean {
  const idx = loresheets.getLoresheetIndex(state, uid);
  if (idx !== -1) { return false; } // Already opened
  const cost: ICost = getCostOpenLoresheet(state, uid, openCost);
  return _canHandleCost(state, cost);
}

export function getCostOpenLoresheet(state: IStoreState, uid: string, cost: number): ICost {
  const idx = state.get('discounts')
    .findIndex((r: IDiscount) => {
      return ( r.type === constants.DISCOUNT_LORESHEET )
      && (r.uids.findIndex((id: string) => (id === uid)) >= 0);
    });
  return _handleDiscount(state, idx, cost);
}

export function canBuyOptionLoresheet(state: IStoreState, lsUid: string, uid: string, buyCost: number): boolean {
  if ( loresheets.getLoresheetIndex(state, lsUid) === -1) { return false; } // LS not open

  const dataOpt = dataLoresheets.optionLS(lsUid, uid);
  const idxOpt = loresheets.getLoresheetOptionIndex(state, lsUid, uid);

  if (idxOpt === -1 && !dataOpt.repeatable) { return false; } // Already bought & not repeatable

  if (dataOpt.prerequisites.filter( // filter checked prereqs
      (p: dataLoresheets.IDataLoresheetOptionPrerequisite) => {
        if (typeof p === 'string') {
          return loresheets.getLoresheetOptionIndex(state, lsUid, p) !== -1;
        } else if ( p.type === 'OR' ) {
          return p.prerequisites
            .filter(
              (prerequisiteUid: string) => (loresheets.getLoresheetOptionIndex(state, lsUid, prerequisiteUid) !== -1)
            ).length > 0;
        } else if ( p.type === 'AND') {
          return p.prerequisites
            .filter(
              (prerequisiteUid: string) => (loresheets.getLoresheetOptionIndex(state, lsUid, prerequisiteUid) !== -1)
            ).length === p.prerequisites.length;
        }
        return false;
      }
    ).length !== dataOpt.prerequisites.length ) { return false; } // prereqs not cleared

  const cost: ICost = getCostBuyOptionLoresheet(state, lsUid, uid, buyCost);
  return _canHandleCost(state, cost);
}

export function getCostBuyOptionLoresheet(state: IStoreState, lsUid: string, uid: string, cost: number): ICost {
  const idx = state.get('discounts')
    .findIndex((r: IDiscount) => {
      return ( r.type === constants.DISCOUNT_LORESHEET_OPTION )
      && (r.uids.findIndex(id => { return id.lsUid === lsUid &&
        id.optUid.findIndex(optUid => (optUid === uid)) >= 0; })
      >= 0);
    });
  return _handleDiscount(state, idx, cost);
}
