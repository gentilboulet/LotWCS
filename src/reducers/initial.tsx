import { makeTypedFactory } from 'typed-immutable-record';
import * as Immutable from 'immutable';
import { IStoreStateJS, IStoreState,
  IStoreSkill, skillFactory,
  IStoreSkillSpecialityJS,
  IStoreLoresheet } from '../types/state';

import { resetToInitialState } from '../actions/initial';
import { IAction } from '../types/actions';
import { IDataSkill } from '../types/skills';
import { skills } from '../data/skills';

export const defaultStateJS: IStoreStateJS = {
  name: 'No Name',
  concept: 'No Concept',
  archetype: '',
  archetypeModified: false,
  rank: '',
  rankValue: -1,
  chi: 0,
  rankModified: false,
  entanglement: 0,
  destiny: 0,
  skills: Immutable.List<IStoreSkill>(skills.map((s: IDataSkill) => {
    return skillFactory({
      name: s.name,
      value: 0,
      specialities: Immutable.List<IStoreSkillSpecialityJS>(s.specialities.map(
        (spe: string) => { return {name: spe, bought: false}; }
      )),
    });
  })),

  loresheets: Immutable.List<IStoreLoresheet>([]),
  reductions: Immutable.List([]),
  bonuses: Immutable.List([]),

  history: Immutable.List<IAction>([resetToInitialState()]),
};

export const initialStateFactory = makeTypedFactory<IStoreStateJS, IStoreState>(defaultStateJS);
