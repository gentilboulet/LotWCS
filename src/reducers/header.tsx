import { IHeaderAction } from '../actions/header';
import { IStoreState } from '../types/state';
import * as constants from '../constants/header';

// Sub Reducers
import { IBonus } from '../types/bonuses';
import { IReduction } from '../types/reductions';
import { applyBonuses } from './bonuses';
import { pushToReductions } from './reductions';
import { pushToHistory } from './history';

export function headerReducer(oldState: IStoreState, action: IHeaderAction): IStoreState {
  switch (action.type) {
    case constants.HEADER_SET_NAME:
      return oldState.withMutations(state => {
        state.set('name', action.name);
        pushToHistory(state, action);
      });
    case constants.HEADER_SET_CONCEPT:
      return oldState.withMutations(state => {
        state.set('concept', action.concept);
        pushToHistory(state, action);
      });
    case constants.HEADER_SET_RANK:
      return oldState.withMutations(state => {
        state.set('rank', action.rank );
        state.set('rankModified', true);
        state.set('rankValue', action.value);
        applyBonuses(state, action.perks as IBonus[]);
        pushToReductions(state, action.perks as IReduction[]);
        pushToHistory(state, action);
      });
    case constants.HEADER_SET_ARCHETYPE:
      return oldState.withMutations(state => {
        state.set('archetype', action.archetype );
        state.set('archetypeModified', true);
        pushToHistory(state, action);
      });
    default:
  }
  return oldState;
}
