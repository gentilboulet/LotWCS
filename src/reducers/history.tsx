import { IAction } from '../types/actions';
import { IStoreState } from '../types/state';
import { IHistoryAction } from '../actions/history';
import * as constants from '../constants/history';
import { globalReducer } from './global';

export function pushToHistory(state: IStoreState, action: IAction): IStoreState {
  return state.updateIn(['history'], (list) => { return list.push(action); });
}

export function replayHistory(state: IStoreState, actions: IAction[]): IStoreState {
  for (let a of actions) { state = globalReducer(state, a); }
  return state;
}

export function historyReducer(oldState: IStoreState, action: IHistoryAction): IStoreState {
  switch (action.type) {
    case constants.HISTORY_DELETE:
      const list: IAction[] = oldState.get('history').toJS();
      return replayHistory(oldState, list.slice(0, action.id));
    default:
      return oldState;
  }
}
