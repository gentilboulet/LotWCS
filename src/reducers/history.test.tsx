import { pushToHistory, replayHistory } from './history';
import { initialStateFactory } from './initial';
import { IAction } from '../types/actions';
import { IStoreState } from '../types/state';

import * as header from '../actions/header';
import { resetToInitialState } from '../actions/initial';
import { historyDelete, IHistoryAction } from '../actions/history';
import { historyReducer } from './history';

const initialState: IStoreState = initialStateFactory();

it('initialState', () => {
  expect( initialState ).toMatchSnapshot();
});

it('pushToHistory', () => {
  expect( pushToHistory(initialState, header.headerSetName('Dummy name'))).toMatchSnapshot();
});

it('replayHistory', () => {
  const actions: IAction[] = [
    resetToInitialState(),
    header.headerSetName('Name'),
    header.headerSetConcept('Concept'),
    resetToInitialState(),
    header.headerSetName('Name2')
  ];
  expect( replayHistory(initialState, actions)).toMatchSnapshot();
});

it('deleteHistory', () => {
  const actions: IAction[] = [
    resetToInitialState(),
    header.headerSetName('SetName'),
    header.headerSetConcept('SetConcept'),
    header.headerSetArchetype('warrior'),
    header.headerSetRank('4th_rank'),
    historyDelete(2)
  ];
  expect( replayHistory(initialState, actions)).toMatchSnapshot();
});

it('JUNK_ACTION', () => {
  const junk = { type: 'JUNK_ACTION' };
  expect( historyReducer(initialState, junk as IHistoryAction )).toMatchSnapshot();
});
