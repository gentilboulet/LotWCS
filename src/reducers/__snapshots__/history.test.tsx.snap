// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing historyReducer should do nothing with a junk action 1`] = `
Object {
  "archetype": "",
  "archetypeModified": false,
  "bonuses": Array [],
  "chi": 0,
  "concept": "No Concept",
  "destiny": 0,
  "discounts": Array [],
  "entanglement": 0,
  "history": Array [
    Object {
      "type": "INITIAL_STATE",
    },
  ],
  "loresheetOptions": Array [],
  "loresheets": Array [],
  "name": "No Name",
  "rank": "",
  "rankModified": false,
  "rankValue": -1,
  "skillSpecialities": Array [],
  "skills": Array [
    Object {
      "name": "Awareness",
      "value": 0,
    },
    Object {
      "name": "Confidence",
      "value": 0,
    },
    Object {
      "name": "Crafting",
      "value": 0,
    },
    Object {
      "name": "Finesse",
      "value": 0,
    },
    Object {
      "name": "Hardiness",
      "value": 0,
    },
    Object {
      "name": "Inspire",
      "value": 0,
    },
    Object {
      "name": "Learning",
      "value": 0,
    },
    Object {
      "name": "Medicine",
      "value": 0,
    },
    Object {
      "name": "Might",
      "value": 0,
    },
    Object {
      "name": "Perform",
      "value": 0,
    },
    Object {
      "name": "Politics",
      "value": 0,
    },
    Object {
      "name": "Ride",
      "value": 0,
    },
    Object {
      "name": "Stealth",
      "value": 0,
    },
    Object {
      "name": "Survival",
      "value": 0,
    },
    Object {
      "name": "Tactics",
      "value": 0,
    },
    Object {
      "name": "Wu Wei",
      "value": 0,
    },
  ],
}
`;

exports[`Testing historyReducer should receive an HISTORY_DELETE action 1`] = `
Object {
  "archetype": "warrior",
  "archetypeModified": true,
  "bonuses": Array [],
  "chi": 10,
  "concept": "Dread pirate Roberts",
  "destiny": 20,
  "discounts": Array [
    Object {
      "skills": Array [
        "Awareness",
        "Confidence",
        "Crafting",
        "Finesse",
        "Hardiness",
        "Inspire",
        "Learning",
        "Medicine",
        "Might",
        "Perform",
        "Politics",
        "Ride",
        "Stealth",
        "Survival",
        "Tactics",
        "Wu Wei",
      ],
      "type": "DISCOUNT_SKILL",
      "value": 20,
    },
  ],
  "entanglement": 0,
  "history": Array [
    Object {
      "type": "INITIAL_STATE",
    },
    Object {
      "name": "Roberts",
      "type": "HEADER_SET_NAME",
    },
    Object {
      "concept": "Dread pirate Roberts",
      "type": "HEADER_SET_CONCEPT",
    },
    Object {
      "archetype": "warrior",
      "type": "HEADER_SET_ARCHETYPE",
    },
    Object {
      "perks": Array [
        Object {
          "type": "BONUS_DESTINY",
          "value": 20,
        },
        Object {
          "skills": Array [
            "Awareness",
            "Confidence",
            "Crafting",
            "Finesse",
            "Hardiness",
            "Inspire",
            "Learning",
            "Medicine",
            "Might",
            "Perform",
            "Politics",
            "Ride",
            "Stealth",
            "Survival",
            "Tactics",
            "Wu Wei",
          ],
          "type": "DISCOUNT_SKILL",
          "value": 20,
        },
        Object {
          "type": "BONUS_STARTING_CHI",
          "value": 10,
        },
      ],
      "rank": "4th_rank",
      "type": "HEADER_SET_RANK",
      "value": 2,
    },
  ],
  "loresheetOptions": Array [],
  "loresheets": Array [],
  "name": "Roberts",
  "rank": "4th_rank",
  "rankModified": true,
  "rankValue": 2,
  "skillSpecialities": Array [],
  "skills": Array [
    Object {
      "name": "Awareness",
      "value": 0,
    },
    Object {
      "name": "Confidence",
      "value": 0,
    },
    Object {
      "name": "Crafting",
      "value": 0,
    },
    Object {
      "name": "Finesse",
      "value": 0,
    },
    Object {
      "name": "Hardiness",
      "value": 0,
    },
    Object {
      "name": "Inspire",
      "value": 0,
    },
    Object {
      "name": "Learning",
      "value": 0,
    },
    Object {
      "name": "Medicine",
      "value": 0,
    },
    Object {
      "name": "Might",
      "value": 0,
    },
    Object {
      "name": "Perform",
      "value": 0,
    },
    Object {
      "name": "Politics",
      "value": 0,
    },
    Object {
      "name": "Ride",
      "value": 0,
    },
    Object {
      "name": "Stealth",
      "value": 0,
    },
    Object {
      "name": "Survival",
      "value": 0,
    },
    Object {
      "name": "Tactics",
      "value": 0,
    },
    Object {
      "name": "Wu Wei",
      "value": 0,
    },
  ],
}
`;

exports[`Testing historyReducer should receive an HISTORY_DELETE action 2`] = `
Object {
  "archetype": "",
  "archetypeModified": false,
  "bonuses": Array [],
  "chi": 0,
  "concept": "No Concept",
  "destiny": 0,
  "discounts": Array [],
  "entanglement": 0,
  "history": Array [
    Object {
      "type": "INITIAL_STATE",
    },
    Object {
      "name": "Roberts",
      "type": "HEADER_SET_NAME",
    },
  ],
  "loresheetOptions": Array [],
  "loresheets": Array [],
  "name": "Roberts",
  "rank": "",
  "rankModified": false,
  "rankValue": -1,
  "skillSpecialities": Array [],
  "skills": Array [
    Object {
      "name": "Awareness",
      "value": 0,
    },
    Object {
      "name": "Confidence",
      "value": 0,
    },
    Object {
      "name": "Crafting",
      "value": 0,
    },
    Object {
      "name": "Finesse",
      "value": 0,
    },
    Object {
      "name": "Hardiness",
      "value": 0,
    },
    Object {
      "name": "Inspire",
      "value": 0,
    },
    Object {
      "name": "Learning",
      "value": 0,
    },
    Object {
      "name": "Medicine",
      "value": 0,
    },
    Object {
      "name": "Might",
      "value": 0,
    },
    Object {
      "name": "Perform",
      "value": 0,
    },
    Object {
      "name": "Politics",
      "value": 0,
    },
    Object {
      "name": "Ride",
      "value": 0,
    },
    Object {
      "name": "Stealth",
      "value": 0,
    },
    Object {
      "name": "Survival",
      "value": 0,
    },
    Object {
      "name": "Tactics",
      "value": 0,
    },
    Object {
      "name": "Wu Wei",
      "value": 0,
    },
  ],
}
`;

exports[`Testing replayHistory should replay an history of actions 1`] = `
Object {
  "archetype": "",
  "archetypeModified": false,
  "bonuses": Array [],
  "chi": 0,
  "concept": "No Concept",
  "destiny": 0,
  "discounts": Array [],
  "entanglement": 0,
  "history": Array [
    Object {
      "type": "INITIAL_STATE",
    },
    Object {
      "name": "John",
      "type": "HEADER_SET_NAME",
    },
  ],
  "loresheetOptions": Array [],
  "loresheets": Array [],
  "name": "John",
  "rank": "",
  "rankModified": false,
  "rankValue": -1,
  "skillSpecialities": Array [],
  "skills": Array [
    Object {
      "name": "Awareness",
      "value": 0,
    },
    Object {
      "name": "Confidence",
      "value": 0,
    },
    Object {
      "name": "Crafting",
      "value": 0,
    },
    Object {
      "name": "Finesse",
      "value": 0,
    },
    Object {
      "name": "Hardiness",
      "value": 0,
    },
    Object {
      "name": "Inspire",
      "value": 0,
    },
    Object {
      "name": "Learning",
      "value": 0,
    },
    Object {
      "name": "Medicine",
      "value": 0,
    },
    Object {
      "name": "Might",
      "value": 0,
    },
    Object {
      "name": "Perform",
      "value": 0,
    },
    Object {
      "name": "Politics",
      "value": 0,
    },
    Object {
      "name": "Ride",
      "value": 0,
    },
    Object {
      "name": "Stealth",
      "value": 0,
    },
    Object {
      "name": "Survival",
      "value": 0,
    },
    Object {
      "name": "Tactics",
      "value": 0,
    },
    Object {
      "name": "Wu Wei",
      "value": 0,
    },
  ],
}
`;
